src/map/skill.cpp

```diff
@@ -14512,21 +14513,24 @@ int32 skill_castend_pos2(struct block_list* src, int32 x, int32 y, uint16 skill_
 	case SA_VOLCANO:
 	case SA_DELUGE:
 	case SA_VIOLENTGALE:
-	{	//Does not consumes if the skill is already active. [Skotlex]
-		std::shared_ptr<s_skill_unit_group> sg2;
-		if ((sg2= skill_locate_element_field(src)) != nullptr && ( sg2->skill_id == SA_VOLCANO || sg2->skill_id == SA_DELUGE || sg2->skill_id == SA_VIOLENTGALE ))
-		{
-			if (sg2->limit - DIFF_TICK(gettick(), sg2->tick) > 0)
-			{
-				skill_unitsetting(src,skill_id,skill_lv,x,y,0);
-				return 0; // not to consume items
+		// Does not consumes if the skill is already active. [Skotlex]
+		if (std::shared_ptr<s_skill_unit_group> sg2 = skill_locate_element_field(src); sg2 != nullptr) {
+			switch( sg2->skill_id ) {
+				case SA_VOLCANO:
+				case SA_DELUGE:
+				case SA_VIOLENTGALE:
+					if (sg2->limit - DIFF_TICK(gettick(), sg2->tick) > 0) {
+						skill_unitsetting(src,skill_id,skill_lv,x,y,0);
```

Comment:
Do we ensure that the duration stays the same?

Also I think the consume check only applies to PC (but maybe that's already handled somewhere else in the code).

Reply:
Not sure what you mean, but the 3 skills have the same duration at all levels (5min)

Reply:
What I mean is that if you have 30 seconds left of those 5 minutes and then you recast, not consuming a gemstone, shouldn't the remaining duration still be 30 seconds?

Reply:
Yeah, that's exactly how it should work